plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,553 @@
+							    basic_data_size +
+								fu_chunk_get_data_sz(chk));
+		fu_struct_algoltek_cmd_transfer_pkt_set_cmd(part_isp_data, FU_ALGOLTEK_CMD_ISP);
+		fu_struct_algoltek_cmd_transfer_pkt_set_address(part_isp_data,
+								fu_chunk_get_address(chk));
+		if (!fu_struct_algoltek_cmd_transfer_pkt_set_data(part_isp_data,
+								  fu_chunk_get_data(chk),
+								  fu_chunk_get_data_sz(chk),
+								  error)) {
+			g_prefix_error(error, "assign isp data failure: ");
+			return FALSE;
+		}
+		fu_struct_algoltek_cmd_transfer_pkt_set_checksum(
+		    part_isp_data,
+		    ~fu_sum8(part_isp_data->data, part_isp_data->len) + 1);
+
+		if (!algoltek_device_ctrl_transfer(self,
+						   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+						   FU_ALGOLTEK_CMD_ISP,
+						   0,
+						   0,
+						   part_isp_data,
+						   part_isp_data->data[0],
+						   error))
+			return FALSE;
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_bot(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_address_pkt_new();
+
+	fu_struct_algoltek_cmd_address_pkt_set_len(buf, 5);
+	fu_struct_algoltek_cmd_address_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_BOT);
+	fu_struct_algoltek_cmd_address_pkt_set_address(buf, address);
+	fu_struct_algoltek_cmd_address_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_BOT,
+					   0,
+					   0,
+					   buf,
+					   buf->data[0],
+					   error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+algoltek_device_ers(FuAlgoltekUsbDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) buf = fu_struct_algoltek_cmd_address_pkt_new();
+
+	fu_struct_algoltek_cmd_address_pkt_set_len(buf, 3);
+	fu_struct_algoltek_cmd_address_pkt_set_cmd(buf, FU_ALGOLTEK_CMD_ERS);
+	fu_struct_algoltek_cmd_address_pkt_set_checksum(buf, ~fu_sum8(buf->data, buf->len) + 1);
+
+	if (!algoltek_device_ctrl_transfer(self,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   FU_ALGOLTEK_CMD_ERS,
+					   0,
+					   0,
+					   buf,
+					   buf->len,
+					   error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_device_status_check(FuDevice *self, gpointer user_data, GError **error)
+{
+	guint16 update_status;
+	g_autoptr(GByteArray) update_status_array = g_byte_array_new();
+
+	update_status_array =
+	    algoltek_device_rdr(FU_ALGOLTEK_USB_DEVICE(self), AG_UPDATE_STATUS, error);
+	if (update_status_array == NULL)
+		return FALSE;
+
+	update_status = fu_algoltek_readout_value(update_status_array, error);
+	if (update_status == AG_UPDATE_READOUT_FAIL)
+		return FALSE;
+
+	switch (update_status) {
+	case AG_UPDATE_PASS:
+		break;
+	case AG_UPDATE_FAIL:
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "update procedure is failed.");
+		return FALSE;
+	default:
+		return FALSE;
```

Comment:
You need to set error in this case too. Maybe combine with the above case?

Reply:
Yes, I combine with the above case.